doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst

```diff
@@ -69,18 +69,127 @@ Bonding
 connected through the IPC instance:
 
-1. The domain (or CPU) writes a magic number to its ``tx-region`` of the shared
-memory.
+#. The domain (or CPU) writes a magic number to its ``tx-region`` of the shared
+   memory.
 #. It then sends a signal to the other domain or CPU, informing that the data
-has been written. Sending the signal to the other domain or CPU is repeated
-with timeout specified by
-:kconfig:option:`CONFIG_IPC_SERVICE_ICMSG_BOND_NOTIFY_REPEAT_TO_MS` option.
+   has been written. Sending the signal to the other domain or CPU is repeated
+   with timeout specified by
+   :kconfig:option:`CONFIG_IPC_SERVICE_ICMSG_BOND_NOTIFY_REPEAT_TO_MS` option.
 #. When the signal from the other domain or CPU is received, the magic number
-is read from ``rx-region``. If it is correct, the bonding process is finished
-and the backend informs the application by calling
-:c:member:`ipc_service_cb.bound` callback.
+   is read from ``rx-region``. If it is correct, the bonding process is finished
+   and the backend informs the application by calling
+   :c:member:`ipc_service_cb.bound` callback.
 
 Samples
 =======
 
  - :zephyr:code-sample:`ipc-icmsg`
+
+Detailed Protocol Specification
+===============================
+
+The ICMsg uses two shared memory regions and two MBOX channels.
+The region and channel pair are used to transfer messages in one direction.
+The other pair is symmetric and transfers messages in the opposite direction.
+For this reason, the specification below focuses on one such pair.
+The other pair is identical.
+
+The ICMsg provides just one endpoint per instance.
+
+Shared Memory Region Organization
+---------------------------------
+
+If data caching is enabled, the shared memory region provided to ICMsg must be
+aligned according to the cache requirement.
+If cache is not enabled, the required alignment is 4 bytes.
+
+The shared memory region is entirely used by a single FIFO.
+It contains read and write indexes followed by the data buffer.
+Detailed structure is contained in the following table.
+
+.. list-table::
+   :header-rows: 1
+
+   * - Field name
+     - Size (bytes)
+     - Byte order
+     - Description
+   * - ``rd_idx``
+     - 4
+     - little‑endian
+     - Index of the first incoming byte in the ``data`` field.
+   * - ``padding``
+     - depends on cache alignment
+     - n/a
+     - Padding added to align ``wr_idx`` to the cache alignment.
+   * - ``wr_idx``
+     - 4
+     - little‑endian
+     - Index of the byte after the last incoming byte in the ``data`` field.
+   * - ``data``
+     - everything to the end of the region
+     - n/a
+     - Circular buffer containing actual bytes to transfer.
+
+This is usual FIFO with a circular buffer, so:
+
+* indexes (``rd_idx`` and ``wr_idx``) are wrapped around when they reach the end of the ``data`` buffer,
+* when ``rd_idx == wr_idx``, the FIFO is empty,
+* it has one byte less capacity than the ``data`` buffer length.
```

Comment:
```suggestion
This is usual FIFO with a circular buffer, so indexes (``rd_idx`` and ``wr_idx``) are wrapped around under the following conditions:

* When they reach the end of the ``data`` buffer.
* When ``rd_idx == wr_idx``, the FIFO is empty,
* When it has one byte less capacity than the ``data`` buffer length.
```

Please check if the meaning has changed here.

Reply:
It is different meaning. I was trying to say that this is usual FIFO, but such statement is not strict enough, so I added those 3 points characterizing it for clarification. Each of them is independent, and you joined them which is wrong. Maybe rewrite the points as follows?
```
This is usual FIFO with a circular buffer:

* The indexes (``rd_idx`` and ``wr_idx``) are wrapped around when they reach the end of the ``data`` buffer.
* The FIFO is empty if ``rd_idx == wr_idx``.
* The FIFO has one byte less capacity than the ``data`` buffer length.
```

Reply:
Agree.